File: C++/12-Tries(PrefixTrees)/Problems/2-WordSearch-II/WordSearch-IIPro.js
/**
 * Word Search II
 *
 * Given an m x n board of characters and a list of strings words, return all words on the board.
 *
 * A word can be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring.
 * The same letter cell may not be used more than once in a word.
 *
 *
 * Approach 1: Basic Trie Search (Conceptual)
 * - Build Trie from words, Search paths in board, Uses simple DFS
 * Approach 2: Optimized Trie Search
 * - Uses Trie, DFS, Marks visited cells, Backtracks, Prevents duplicates
 * Approach 3: Trie Search with Early Termination
 * - Uses Trie, DFS, Visited set, Prunes Trie branches without matches
 * Approach 4: Trie Search with Set for Result
 * - Uses Trie, DFS, Set for results, Optimized board traversal
 * Approach 5: Aho-Korasick Automata (Advanced)
 * - Builds Aho-Korasick automaton, Searches, Efficient for many words
 *
 */

// --- Helper Functions ---

/**
 * Builds a Trie from a list of words.
 * @param {string[]} words - The list of words.
 * @returns {object} The Trie data structure.
 */
function buildTrie(words) {
    const trie = {};
    for (const word of words) {
        let node = trie;
        for (const char of word) {
            if (!node[char]) {
                node[char] = {};
            }
            node = node[char];
        }
        node["$"] = word; // Mark the end of a word with a special character
    }
    return trie;
}

/**
 * Depth-First Search to find words in the board.
 * @param {character[][]} board - The game board.
 * @param {number} row - The current row index.
 * @param {number} col - The current column index.
 * @param {object} node - The current node in the Trie.
 * @param {Set<string>} foundWords - Set to store found words.
 */
function dfs(board, row, col, node, foundWords) {
    // (Implementation varies slightly across approaches)
}

// --- Approach 1: Basic Trie Search (Conceptual) ---
function findWords1(board, words) {
    const trie = buildTrie(words);
    const foundWords = new Set();

    function dfs(board, row, col, node, foundWords) {
        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length) {
            return;
        }
        const char = board[row][col];
        if (!node[char]) {
            return;
        }
        node = node[char];
        if (node["$"]) {
            foundWords.add(node["$"]);
        }

        // Explore adjacent cells
        dfs(board, row + 1, col, node, foundWords);
        dfs(board, row - 1, col, node, foundWords);
        dfs(board, row, col + 1, node, foundWords);
        dfs(board, row, col - 1, node, foundWords);
    }

    for (let r = 0; r < board.length; r++) {
        for (let c = 0; c < board[0].length; c++) {
            dfs(board, r, c, trie, foundWords);
        }
    }
    return Array.from(foundWords);
}

// --- Approach 2: Optimized Trie Search ---
function findWords2(board, words) {
    const trie = buildTrie(words);
    const foundWords = new Set();
    const rows = board.length;
    const cols = board[0].length;

    function dfs(board, row, col, node, foundWords, visited) {
        if (row < 0 || row >= rows || col < 0 || col >= cols || visited[row][col]) {
            return;
        }

        const char = board[row][col];
        if (!node[char]) {
            return;
        }

        visited[row][col] = true;
        node = node[char];

        if (node["$"]) {
            foundWords.add(node["$"]);
        }

        dfs(board, row + 1, col, node, foundWords, visited);
        dfs(board, row - 1, col, node, foundWords, visited);
        dfs(board, row, col + 1, node, foundWords, visited);
        dfs(board, row, col - 1, node, foundWords, visited);

        visited[row][col] = false; // Backtrack: unmark the cell
    }

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            const visited = Array(rows).fill(null).map(() => Array(cols).fill(false));
            dfs(board, r, c, trie, foundWords, visited);
        }
    }
    return Array.from(foundWords);
}

// --- Approach 3: Trie Search with Early Termination ---
function findWords3(board, words) {
    const trie = buildTrie(words);
    const foundWords = new Set();
    const rows = board.length;
    const cols = board[0].length;

    function dfs(board, row, col, node, foundWords, visited) {
        if (row < 0 || row >= rows || col < 0 || col >= cols || visited[row][col]) {
            return;
        }

        const char = board[row][col];
        if (!node[char]) {
            return;
        }

        visited[row][col] = true;
        const nextNode = node[char]; // Store nextNode
        if (nextNode["$"]) {
            foundWords.add(nextNode["$"]);
        }

        dfs(board, row + 1, col, nextNode, foundWords, visited);
        dfs(board, row - 1, col, nextNode, foundWords, visited);
        dfs(board, row, col + 1, nextNode, foundWords, visited);
        dfs(board, row, col - 1, nextNode, foundWords, visited);

        visited[row][col] = false;
    }

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            const visited = Array(rows).fill(null).map(() => Array(cols).fill(false));
            dfs(board, r, c, trie, foundWords, visited);
        }
    }
    return Array.from(foundWords);
}

// --- Approach 4: Trie Search with Set for Result ---
function findWords4(board, words) {
    const trie = buildTrie(words);
    const foundWords = new Set();
    const rows = board.length;
    const cols = board[0].length;

    function dfs(board, row, col, node, foundWords, visited) {
        if (row < 0 || row >= rows || col < 0 || col >= cols || visited[row][col]) {
            return;
        }

        const char = board[row][col];
        if (!node[char]) {
            return;
        }

        visited[row][col] = true;
        const nextNode = node[char];
        if (nextNode["$"]) {
            foundWords.add(nextNode["$"]);
        }

        dfs(board, row + 1, col, nextNode, foundWords, visited);
        dfs(board, row - 1, col, nextNode, foundWords, visited);
        dfs(board, row, col + 1, nextNode, foundWords, visited);
        dfs(board, row, col - 1, nextNode, foundWords, visited);

        visited[row][col] = false;
    }

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
             const visited = Array(rows).fill(null).map(() => Array(cols).fill(false));
            dfs(board, r, c, trie, foundWords, visited);
        }
    }
    return Array.from(foundWords);
}

// --- Approach 5: Aho-Korasick Automata (Advanced) ---
function findWords5(board, words) {
    // Aho-Korasick is complex.  This is a simplified conceptual outline.
    //  Real implementation requires a queue, and careful state management.
    //  This approach is generally VERY efficient for a large number of words.

    const rows = board.length;
    const cols = board[0].length;
    const foundWords = new Set();

    // 1. Build Trie (as before, but conceptually part of Aho-Korasick)
    const trie = buildTrie(words);

    // 2. Build Failure Function (the core of Aho-Korasick - skipped for brevity)
    //    -  This would involve a BFS on the Trie.
    //    -  Each node gets a 'failure' link to the longest proper suffix that is also a prefix in the Trie.

    // 3. Search the board using the automaton (Conceptual)
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            let currentState = trie; // Start at root
            let currentRow = r;
            let currentCol = c;

            function searchFromCell(row, col, state) {

                if (row < 0 || row >= rows || col < 0 || col >= cols) return;
                const char = board[row][col];

                 if (!state[char]) {
                    return; // Dead end
                }
                state = state[char];

                if (state["$"]) {
                    foundWords.add(state["$"]);
                }
                searchFromCell(row + 1, col, state);
                searchFromCell(row - 1, col, state);
                searchFromCell(row, col+1, state);
                searchFromCell(row, col-1, state);
            }
            searchFromCell(currentRow, currentCol, currentState);
        }
    }
    return Array.from(foundWords);
}

// --- Test Cases ---
const board1 = [
    ["o", "a", "a", "n"],
    ["e", "t", "a", "e"],
    ["i", "h", "k", "r"],
    ["i", "f", "l", "v"],
];
const words1 = ["oath", "pea", "eat", "rain"];

const board2 = [["a", "b"], ["c", "d"]];
const words2 = ["ab", "cb", "ad", "bd", "ac", "ca", "da", "bc", "db", "adcb", "dabc"];

const board3 = [["a"]];
const words3 = ["a"];

const board4 = [
    ["a","a","a","a"],
    ["a","a","a","a"],
    ["a","a","a","a"]
];
const words4 = ["aaaaaaaaaaaa","aaaaaaaaaaaaa","aaaaaaaaaa"];

const board5 = [
    ["b","a","a","b","a","b"],
    ["a","b","a","a","a","a"],
    ["a","b","b","a","b","a"],
    ["a","a","a","a","a","b"],
    ["b","a","b","a","b","a"],
    ["a","b","a","b","b","b"]
];
const words5 = ["abbbab","baab","baa"];


console.log("--- Test Case 1 ---");
console.log("Input Board:", board1);
console.log("Input Words:", words1);
console.log("Approach 1:", findWords1(board1, words1));
console.log("Approach 2:", findWords2(board1, words1));
console.log("Approach 3:", findWords3(board1, words1));
console.log("Approach 4:", findWords4(board1, words1));
console.log("Approach 5:", findWords5(board1, words1));

console.log("\n--- Test Case 2 ---");
console.log("Input Board:", board2);
console.log("Input Words:", words2);
console.log("Approach 1:", findWords1(board2, words2));
console.log("Approach 2:", findWords2(board2, words2));
console.log("Approach 3:", findWords3(board2, words2));
console.log("Approach 4:", findWords4(board2, words2));
console.log("Approach 5:", findWords5(board2, words2));

console.log("\n--- Test Case 3 ---");
console.log("Input Board:", board3);
console.log("Input Words:", words3);
console.log("Approach 1:", findWords1(board3, words3));
console.log("Approach 2:", findWords2(board3, words3));
console.log("Approach 3:", findWords3(board3, words3));
console.log("Approach 4:", findWords4(board3, words3));
console.log("Approach 5:", findWords5(board3, words3));

console.log("\n--- Test Case 4 ---");
console.log("Input Board:", board4);
console.log("Input Words:", words4);
console.log("Approach 1:", findWords1(board4, words4));
console.log("Approach 2:", findWords2(board4, words4));
console.log("Approach 3:", findWords3(board4, words4));
console.log("Approach 4:", findWords4(board4, words4));
console.log("Approach 5:", findWords5(board4, words4));

console.log("\n--- Test Case 5 ---");
console.log("Input Board:", board5);
console.log("Input Words:", words5);
console.log("Approach 1:", findWords1(board5, words5));
console.log("Approach 2:", findWords2(board5, words5));
console.log("Approach 3:", findWords3(board5, words5));
console.log("Approach 4:", findWords4(board5, words5));
console.log("Approach 5:", findWords5(board5, words5));

File: Java/12-Tries(PrefixTrees)/Problems/2-WordSearch-II/WordSearch-IIPro.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


// Class to represent the Trie data structure
class Trie {
    public static void main(String[] args) {
        WordSearchII solution = new WordSearchII();
        char[][] board = {
                {'o', 'a', 'a', 'n'},
                {'e', 't', 'a', 'e'},
                {'i', 'h', 'k', 'r'},
                {'i', 'f', 'l', 'v'}
        };
        String[] words = {"oath", "pea", "eat", "rain"};

        System.out.println("Approach 1: " + solution.findWords1(board, words));
        System.out.println("Approach 2: " + solution.findWords2(board, words));
        System.out.println("Approach 3: " + solution.findWords3(board, words));
        System.out.println("Approach 4: " + solution.findWords4(board, words));
        System.out.println("Approach 5: " + solution.findWords5(board, words));
    }

    TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    // Method to insert a word into the Trie
    public void insert(String word) {
        TrieNode current = root;
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (current.children[index] == null) {
                current.children[index] = new TrieNode();
            }
            current = current.children[index];
        }
        current.isEndOfWord = true;
        current.word = word; // Store the complete word at the end node
    }
}

public class WordSearchII {

    // 1.  Backtracking with Trie - Basic Implementation
    //     -   Uses a Trie to store words and backtracking to search the board.
    //     -   Good for understanding the basic approach.
    //     -   Time Complexity: O(M * N * 4^L), where M is rows, N is cols, L is longest word.
    //     -   Space Complexity: O(W), where W is the total number of characters in words.
    public List<String> findWords1(char[][] board, String[] words) {
        Trie trie = new Trie();
        for (String word : words) {
            trie.insert(word);
        }
        Set<String> result = new HashSet<>();
        int rows = board.length;
        int cols = board[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                search1(board, i, j, trie.root, result);
            }
        }
        return new ArrayList<>(result);
    }

    private void search1(char[][] board, int row, int col, TrieNode node, Set<String> result) {
        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || board[row][col] == '#' || node == null) {
            return;
        }

        char c = board[row][col];
        int index = c - 'a';
        if (node.children[index] == null) {
            return;
        }

        node = node.children[index]; // Move to the next node in the trie
        if (node.isEndOfWord) {
            result.add(node.word);
        }

        // Mark the current cell as visited
        board[row][col] = '#';
        // Explore all 4 directions
        search1(board, row + 1, col, node, result);
        search1(board, row - 1, col, node, result);
        search1(board, row, col + 1, node, result);
        search1(board, row, col - 1, node, result);
        // Backtrack: Restore the original character
        board[row][col] = c;
    }

    // 2. Backtracking with Trie + Optimization 1: Early Stop
    //    - Optimized version of approach 1.  Stops searching a path
    //      as soon as it's clear no word can be formed.
    //    - Prunes unnecessary search paths, improving performance.
    //    - Time Complexity: O(M * N * 4^L), but performs better in average case.
    //    - Space Complexity: O(W)
    public List<String> findWords2(char[][] board, String[] words) {
        Trie trie = new Trie();
        for (String word : words) {
            trie.insert(word);
        }
        Set<String> result = new HashSet<>();
        int rows = board.length;
        int cols = board[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                search2(board, i, j, trie.root, result);
            }
        }
        return new ArrayList<>(result);
    }

    private void search2(char[][] board, int row, int col, TrieNode node, Set<String> result) {
        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || board[row][col] == '#' || node == null) {
            return;
        }

        char c = board[row][col];
        int index = c - 'a';
        if (node.children[index] == null) {
            return; // Stop if there's no matching prefix in Trie
        }

        node = node.children[index];
        if (node.isEndOfWord) {
            result.add(node.word);
            node.isEndOfWord = false; // Avoid duplicates, IMPORTANT for optimization
        }

        board[row][col] = '#';
        search2(board, row + 1, col, node, result);
        search2(board, row - 1, col, node, result);
        search2(board, row, col + 1, node, result);
        search2(board, row, col - 1, node, result);
        board[row][col] = c;
    }

    // 3. Backtracking with Trie + Optimization 2: Trie Node Modification
    //    - Further optimization of approach 2.  Instead of just setting
    //      isEndOfWord to false, we remove the word from the Trie node.
    //    - This prevents revisiting already found words and reduces search space.
    //     - Time Complexity: O(M * N * 4^L), but performs significantly better in practice.
    //    - Space Complexity: O(W)
    public List<String> findWords3(char[][] board, String[] words) {
        Trie trie = new Trie();
        for (String word : words) {
            trie.insert(word);
        }
        Set<String> result = new HashSet<>();
        int rows = board.length;
        int cols = board[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                search3(board, i, j, trie, trie.root, result);
            }
        }
        return new ArrayList<>(result);
    }

    private void search3(char[][] board, int row, int col, Trie trie, TrieNode node, Set<String> result) {
        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || board[row][col] == '#' || node == null) {
            return;
        }

        char c = board[row][col];
        int index = c - 'a';
        if (node.children[index] == null) {
            return;
        }

        TrieNode nextNode = node.children[index];
        if (nextNode.isEndOfWord) {
            result.add(nextNode.word);
            // Remove word from Trie to avoid duplicates and further searching
            nextNode.isEndOfWord = false;
            nextNode.word = null; // Important:  Clear the word!

        }

        board[row][col] = '#';
        search3(board, row + 1, col, trie, nextNode, result);
        search3(board, row - 1, col, trie, nextNode, result);
        search3(board, row, col + 1, trie, nextNode, result);
        search3(board, row, col - 1, trie, nextNode, result);
        board[row][col] = c;
    }

    // 4. Backtracking with Trie + Optimization 3:  Check for Empty Trie Node
    //    -  The most optimized version.  After finding a word, if a Trie node
    //       has no more children, we can effectively prune that node from the
    //       search space.
    //    -  This can significantly reduce the search space for large inputs.
    //    - Time Complexity: O(M * N * 4^L),  best practical performance.
    //    - Space Complexity: O(W)
    public List<String> findWords4(char[][] board, String[] words) {
        Trie trie = new Trie();
        for (String word : words) {
            trie.insert(word);
        }
        Set<String> result = new HashSet<>();
        int rows = board.length;
        int cols = board[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                search4(board, i, j, trie, trie.root, result);
            }
        }
        return new ArrayList<>(result);
    }

    private void search4(char[][] board, int row, int col, Trie trie, TrieNode node, Set<String> result) {
        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || board[row][col] == '#' || node == null) {
            return;
        }

        char c = board[row][col];
        int index = c - 'a';
        if (node.children[index] == null) {
            return;
        }

        TrieNode nextNode = node.children[index];
        if (nextNode.isEndOfWord) {
            result.add(nextNode.word);
            nextNode.isEndOfWord = false;
            nextNode.word = null;
        }

        board[row][col] = '#';
        search4(board, row + 1, col, trie, nextNode, result);
        search4(board, row - 1, col, trie, nextNode, result);
        search4(board, row, col + 1, trie, nextNode, result);
        search4(board, row, col - 1, trie, nextNode, result);
        board[row][col] = c;

        // After exploring all paths from the current node, check if it has become empty.
        boolean hasChildren = false;
        for (int i = 0; i < 26; i++) {
            if (nextNode.children[i] != null) {
                hasChildren = true;
                break;
            }
        }
        if (!hasChildren) {
            // If the node has no children, remove it from its parent.
            node.children[index] = null;
        }
    }

    // 5.  Iterative Approach using Depth-First Search (DFS) with Trie
    //     -   Implements an iterative DFS using a stack to avoid recursion.
    //     -   Useful for very large boards where recursion might lead to stack overflow.
    //     -   Time Complexity: O(M * N * 4^L)
    //     -   Space Complexity: O(W + M * N),  W for Trie, M*N for stack in worst case.
    public List<String> findWords5(char[][] board, String[] words) {
        Trie trie = new Trie();
        for (String word : words) {
            trie.insert(word);
        }
        Set<String> result = new HashSet<>();
        int rows = board.length;
        int cols = board[0].length;
        // Use a Stack to simulate DFS
        java.util.Stack<State> stack = new java.util.Stack<>();

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                stack.push(new State(i, j, trie.root)); // Start DFS from each cell
                while (!stack.isEmpty()) {
                    State currentState = stack.pop();
                    int row = currentState.row;
                    int col = currentState.col;
                    TrieNode node = currentState.node;

                    if (row < 0 || row >= rows || col < 0 || col >= cols || board[row][col] == '#' || node == null) {
                        continue;
                    }

                    char c = board[row][col];
                    int index = c - 'a';
                    if (node.children[index] == null) {
                        continue;
                    }

                    TrieNode nextNode = node.children[index];
                    if (nextNode.isEndOfWord) {
                        result.add(nextNode.word);
                        nextNode.isEndOfWord = false;
                        nextNode.word = null;
                    }
                    board[row][col] = '#'; // Mark visited

                    stack.push(new State(row + 1, col, nextNode));
                    stack.push(new State(row - 1, col, nextNode));
                    stack.push(new State(row, col + 1, nextNode));
                    stack.push(new State(row, col - 1, nextNode));

                    board[row][col] = c; // Backtrack
                }
            }
        }
        return new ArrayList<>(result);
    }

    // Helper class for iterative DFS
    private static class State {
        int row;
        int col;
        TrieNode node;

        public State(int row, int col, TrieNode node) {
            this.row = row;
            this.col = col;
            this.node = node;
        }
    }
}

// Class to represent a node in the Trie
class TrieNode {
    TrieNode[] children;
    String word; // Stores the complete word if this node represents the end of a word
    boolean isEndOfWord;

    public TrieNode() {
        children = new TrieNode[26]; // For 'a' to 'z'
        word = null;
        isEndOfWord = false;
    }
}

/*
Real-world project implementation scenarios for each approach:

1.  Backtracking with Trie - Basic Implementation:
    -   Suitable for small-scale applications where performance is not critical.
    -   Example: A simple word search game in a mobile app with a limited dictionary.
    -   Educational tool to demonstrate Trie and backtracking concepts.

2.  Backtracking with Trie + Optimization 1: Early Stop:
    -   Applicable in scenarios where some optimization is needed.
    -   Example: Implementing word search in a web application with a moderate-sized dictionary.
    -   Useful when response time is important but not extremely critical.

3.  Backtracking with Trie + Optimization 2: Trie Node Modification:
    -   Good for applications with a larger dictionary and more complex board configurations.
    -   Example: A word search puzzle generator or solver with a substantial word list.
    -   Improves efficiency by avoiding redundant searches.

4.  Backtracking with Trie + Optimization 3: Check for Empty Trie Node:
    -   Best for high-performance applications with very large dictionaries.
    -   Example: A server-side component for a multiplayer word game with a massive word database.
    -   Critical in scenarios where minimizing search space is essential for speed.

5.  Iterative Approach using Depth-First Search (DFS) with Trie:
    -   Essential for applications that need to handle extremely large boards or dictionaries.
    -   Example: A professional word search solver or a text analysis tool processing large datasets.
    -   Prevents stack overflow errors and provides a more robust solution.
*/
